### PR TITLE
[WIP] Multi module build

### PR DESCRIPTION
**Pull request by koppor**
This is a huge step towards https://github.com/JabRef/jabref/issues/110 and prepares https://github.com/JabRef/jabref/pull/3421 somehow.

This enables gradle multi module build.

The dependencies of "logic" were not as clean as I thought. I removed dependencies to GUI or moved packages back to gui (shared, remote).

We should get this in ASAP. 

Open issues:

- [ ] `src/main/gen` should be removed from version control - generated code like jabref.gui/src/main/gen/org/jabref/logic/importer/fileformat/mods/RoleTermDefinition.java is put under version control; is this desired?
- [ ] inherit more from root project avoid duplication of plugins dependencies and tasks (https://docs.gradle.org/current/userguide/multi_project_builds.html)
- [x] GUI does not start: `Cannot load file GroupTree.fxml`
- [x] Install4J does not find the generated JARs.
- [ ] The OO preferences should be fixed (reading/writing not fully supported yet). All other features should work. - will be done later https://github.com/koppor/jabref/issues/304
- [x] `jar` are not uploaded to CircleCI - are they stored somewhere else? http://builds.jabref.org/multi-module-build/
- [x]  Tests not executed on travis. `:org.jabref.gui:junitPlatformTest`  - this works locally though.
- [ ] Tests fail on travis. E.g., `[main] ERROR org.jabref.logic.l10n.Localization - Messages are not initialized before accessing All entries`

### CODE REVIEW

**Comment by tobiasdiez** [comment]
The fxml and css files should lie in the same folder as the view `GroupTreeView`

File: org.jabref.gui/src/main/java/org/jabref/preferences/SavePreferencesFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package org.jabref.preferences;
+
+import org.jabref.logic.bibtex.LatexFieldFormatterPreferences;
+import org.jabref.logic.exporter.SavePreferences;
+import org.jabref.model.bibtexkeypattern.GlobalBibtexKeyPattern;
+import org.jabref.model.metadata.SaveOrderConfig;
+
+import java.nio.charset.Charset;
+
+public class SavePreferencesFactory {
```
Comment (tobiasdiez) [inline org.jabref.gui/src/main/java/org/jabref/preferences/SavePreferencesFactory.java]: You can just move these methods to `JabRefPreferences`.

**Comment by tobiasdiez** [comment]
@koppor Please don't move the fxml and css files from source to resources if this somehow possible! These files function as code (in the sense that you could alternatively define everything in java if you want) and it is very convenient to have them right next to the java classes (controller, view model) since you usually edit them in parallel.

**Comment by koppor** [comment]
`AbstractDialogView.class.getResource("Main.css")` returns `null` with the current multi-module build. I have no clue how to convince the JRE (?) to also search next to the class files. Eventually, I have to manually adapt the gradle file to copy all files not ending with `.java` to the output.

**Comment by koppor** [comment]
Current state: Compiles fine on Windows, but on Linux:

```
FAILURE: Build failed with an exception.

* Where:
Build file '/home/ubuntu/jabref/org.jabref.logic/build.gradle' line: 73

* What went wrong:
Execution failed for task ':org.jabref.logic:processResources'.
> No signature of method: static org.gradle.api.Project.getRootProject() is applicable for argument types: () values: []
```

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # no generated files in version control
-src/main/gen/
+**/src/main/gen/
```
Comment (tobiasdiez) [inline .gitignore]: generated code like jabref.gui/src/main/gen/org/jabref/logic/importer/fileformat/mods/RoleTermDefinition.java is put under version control; is this desired?

Comment (koppor) [reply .gitignore]: That should not happen. We have to remove that code.

I will double check.

**Comment by koppor** [comment]
This will be frozen until https://github.com/JabRef/jabref/pull/3733 is merged. Reason: That PR also does huge refactorings.

**Comment by koppor** [comment]
I close this PR as this is unfinished.

I created https://github.com/JabRef/jabref/pull/3913 as follow up. The small changes are filed as PRs https://github.com/JabRef/jabref/pull/3907, https://github.com/JabRef/jabref/pull/3908, https://github.com/JabRef/jabref/pull/3909, https://github.com/JabRef/jabref/pull/3910, https://github.com/JabRef/jabref/pull/3911, https://github.com/JabRef/jabref/pull/3912

The only thing not followed-up is the gradle multi-module-build (Co-Authored-By @stefan-kolb). I propose to leave it to @florian-beetz to investigate the gradle scripts and see whether they help him to get JabRef ready for Java9.

**Pull request closed without merging**